packages/xsnap/src/xsnap.js

```diff
@@ -340,35 +411,112 @@ export async function xsnap(options) {
   }
 
   /**
-   * @param {string} [description]
+   * @param {string} description
+   * @param {import('@endo/promise-kit').PromiseKit<void>} batonKit
    * @returns {AsyncGenerator<Uint8Array, void, undefined>}
    */
-  async function* makeSnapshot(description = 'unknown') {
-    // TODO: Refactor to use tmpFile rather than tmpName.
-    const tmpSnapPath = await ptmpName({
-      template: `make-snapshot-${safeHintFromDescription(
-        description,
-      )}-XXXXXX.xss`,
-    });
-
+  async function* makeSnapshotInternal(description, batonKit) {
+    const output = new PassThrough({ highWaterMark: 1024 * 1024 });
+    let piped = false;
+    let cleaned = false;
+    let done = Promise.resolve();
+
+    let snapshotReadSize = 0;
+    /** @type {number | undefined} */
+    let snapshotSize;
     try {
-      const result = baton.then(async () => {
-        await messagesToXsnap.next(encoder.encode(`w${tmpSnapPath}`));
-        await runToIdle();
+      /** @type {string} */
+      let snapPath;
+      /** @type {Readable} */
+      let sourceStream;
+
+      const maybePipe = () => {
+        if (!piped && !cleaned) {
+          sourceStream.pipe(output);
+          piped = true;
+        }
+      };
+
+      if (snapshotUseFs) {
+        // TODO: Refactor to use tmpFile rather than tmpName.
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        snapPath = await ptmpName({
+          template: `make-snapshot-${safeHintFromDescription(
+            description,
+          )}-XXXXXX.xss`,
+        });
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        const handle = await fs.open(snapPath, 'w+');
+        // @ts-expect-error 'close' event added in Node 15.4
+        handle.on('close', () => {
+          fs.unlink(snapPath);
+        });
+        sourceStream = handle.createReadStream();
+        finished(output).finally(() => sourceStream.destroy());
+      } else {
+        sourceStream = savedSnapshotsStream;
+        snapPath = '@7';
+        batonKit.promise.then(maybePipe);
+      }
+
+      const cleanup = () => {
+        if (cleaned) return;
+        cleaned = true;
+        sourceStream.unpipe(output);
+        // eslint-disable-next-line no-use-before-define
+        output.off('data', onData);
+        output.end();
+      };
+      const checkDone = () => {
+        if (snapshotSize !== undefined && snapshotReadSize >= snapshotSize) {
+          cleanup();
+        }
+      };
+      const onData = chunk => {
+        snapshotReadSize += chunk.length;
+        checkDone();
+      };
+      output.on('data', onData);
+
+      const result = batonKit.promise.then(async () => {
+        await messagesToXsnap.next(encoder.encode(`w${snapPath}`));
+        const { reply } = await runToIdle();
+        const lengthStr = decoder.decode(reply);
+        if (lengthStr.length) {
+          snapshotSize = Number(lengthStr);
+          maybePipe();
+        } else {
+          snapshotSize = -1;
```

Comment:
What does it mean if we get here? I think it means that the snapshot write failed, and xsnap is trying to tell us there was an error, or maybe xsnap died in the middle of the process.

I guess `snapshotSize = -1` means that `cleanup()` will finish quickly, which is good, but where do we tell the stream (going into `snapStore`?) that the data is bad, and it should not be saved? And how does the caller (who just hands the stream to `snapStore`) learn that it failed? I guess maybe `snapStore.saveSnapshot()` is supposed to throw when it sees the stream finish with an error?

Reply:
The `finally` section checks the read size against this size, and if different throws, which means the consumer will see an error before its stream iteration ends. And yes, if iteration throws, the `snapStore.saveSnapshot()` will not store anything and re-throw.